internal/cmd/update.go

```diff
@@ -0,0 +1,466 @@
+		Name               string `json:"name"`
+	}
+	type release struct {
+		Assets []asset `json:"assets"`
+	}
+	var r release
+
+	resp, err := httpClient.Get(urlString)
+	if err != nil {
+		return "", xerrors.Errorf("query github release url %s: %w", urlString, err)
+	}
+	defer resp.Body.Close()
+
+	err = json.NewDecoder(resp.Body).Decode(&r)
+	if err != nil {
+		return "", xerrors.Errorf("unmarshal github releases api response: %w", err)
+	}
+
+	var assetURLStr string
+	for _, a := range r.Assets {
+		if strings.HasPrefix(a.Name, "coder-cli-"+ostype) {
+			assetURLStr = a.BrowserDownloadURL
+		}
+	}
+
+	if assetURLStr == "" {
+		return "", xerrors.Errorf("could not find release for ostype %s", ostype)
+	}
+
+	return assetURLStr, nil
+}
+
+func extractFromArchive(path string, archive []byte) ([]byte, error) {
+	contentType := http.DetectContentType(archive)
+	switch contentType {
+	case "application/zip":
+		return extractFromZip(path, archive)
+	case "application/x-gzip":
+		return extractFromTgz(path, archive)
+	default:
+		return nil, xerrors.Errorf("unknown archive type: %s", contentType)
+	}
+}
+
+func extractFromZip(path string, archive []byte) ([]byte, error) {
+	zipReader, err := zip.NewReader(bytes.NewReader(archive), int64(len(archive)))
+	if err != nil {
+		return nil, xerrors.Errorf("failed to open zip archive")
+	}
+
+	var zf *zip.File
+	for _, f := range zipReader.File {
+		if f.Name == path {
+			zf = f
+			break
+		}
+	}
+	if zf == nil {
+		return nil, xerrors.Errorf("could not find path %q in zip archive", path)
+	}
+
+	rc, err := zf.Open()
+	if err != nil {
+		return nil, xerrors.Errorf("failed to extract path %q from archive", path)
+	}
+	defer rc.Close()
+
+	var b bytes.Buffer
+	bw := bufio.NewWriter(&b)
+	if _, err := io.Copy(bw, rc); err != nil {
+		return nil, xerrors.Errorf("failed to copy path %q to from archive", path)
+	}
+	return b.Bytes(), nil
+}
+
+func extractFromTgz(path string, archive []byte) ([]byte, error) {
+	zr, err := gzip.NewReader(bytes.NewReader(archive))
+	if err != nil {
+		return nil, xerrors.Errorf("failed to gunzip archive")
+	}
+
+	tr := tar.NewReader(zr)
+
+	var b bytes.Buffer
+	bw := bufio.NewWriter(&b)
+	for {
+		hdr, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, xerrors.Errorf("failed to read tar archive: %w", err)
+		}
+		fi := hdr.FileInfo()
+		if fi.Name() == path && fi.Mode().IsRegular() {
+			_, _ = io.Copy(bw, tr)
+			break
+		}
+	}
+
+	return b.Bytes(), nil
+}
+
+// getCoderConfigURL reads the currently configured coder URL, returning an empty string if not configured.
+func getCoderConfigURL() (*url.URL, error) {
+	urlString, err := config.URL.Read()
+	if err != nil {
+		return nil, err
+	}
+	configuredURL, err := url.Parse(strings.TrimSpace(urlString))
+	if err != nil {
+		return nil, err
+	}
+	return configuredURL, nil
+}
+
+// XXX: coder.Client requires an API key, but we may not be logged into the coder instance for which we
```

Comment:
is this still true with the unauthenticated API?

Reply:
I don't think there's any change here. There's no way I can see to instantiate a `coder.Client` without credentials. In any case, it looks like the client just sniffs the HTTP headers to fetch the version.